extensions/amp-a4a/0.1/legacy-signature-verifier.js

```diff
@@ -0,0 +1,161 @@
+        throw new Error('missing fields in JSON Web Key');
+      }
+      const mod = base64UrlDecodeToBytes(jwk.n);
+      const pubExp = base64UrlDecodeToBytes(jwk.e);
+      const lenMod = lenPrefix(mod);
+      const lenPubExp = lenPrefix(pubExp);
+      const data = new Uint8Array(lenMod.length + lenPubExp.length);
+      data.set(lenMod);
+      data.set(lenPubExp, lenMod.length);
+          // The list of RSA public keys are not under attacker's
+          // control, so a collision would not help.
+      return this.crypto_.sha1(data)
+          .then(digest => ({
+            serviceName,
+            cryptoKey,
+                // Hash is the first 4 bytes of the SHA-1 digest.
+            hash: new Uint8Array(digest, 0, 4),
+          }));
+    });
+  }
+
+  /**
+   * Verifies RSA signature corresponds to the data, given a public key.
+   * @param {!Uint8Array} data the data that was signed.
+   * @param {!Uint8Array} signature the RSA signature.
+   * @param {!PublicKeyInfoDef} publicKeyInfo the RSA public key.
+   * @return {!Promise<!boolean>} whether the signature is valid for
+   *     the public key.
+   */
+  verifySignature(data, signature, publicKeyInfo) {
+    if (!verifyHashVersion(signature, publicKeyInfo)) {
+      return Promise.resolve(false);
+    }
+    // Verify that the data matches the raw RSA signature, using the
+    // public key.
+    // Append the version number to the data.
+    const signedData = new Uint8Array(data.length + 1);
+    signedData.set(data);
+    signedData[data.length] = SIGNATURE_VERSION;
+
+    return /** @type {!Promise<boolean>} */ (this.crypto_.verifyPkcs(
+        publicKeyInfo.cryptoKey, signature.subarray(5), signedData));
+  }
+}
+
+/**
+ * Verifies signature was signed with private key matching public key given.
+ * Does not verify data actually matches signature (use verifySignature).
+ * @param {!Uint8Array} signature the RSA signature.
+ * @param {!PublicKeyInfoDef} publicKeyInfo the RSA public key.
+ * @return {boolean} whether signature was generated using hash.
+ */
+function verifyHashVersion(signature, publicKeyInfo) {
+  // The signature has the following format:
+  // 1-byte version + 4-byte key hash + raw RSA signature where
+  // the raw RSA signature is computed over (data || 1-byte version).
+  // If the hash doesn't match, don't bother checking this key.
+  return signature.length > 5 && signature[0] == SIGNATURE_VERSION &&
+      hashesEqual(signature, publicKeyInfo.hash);
+}
+
+/**
+ * Appends 4-byte endian data's length to the data itself.
+ * @param {!Uint8Array} data
+ * @return {!Uint8Array} the prepended 4-byte endian data's length together with
+ *     the data itself.
+ */
+function lenPrefix(data) {
+  const res = new Uint8Array(4 + data.length);
+  res[0] = (data.length >> 24) & 0xff;
+  res[1] = (data.length >> 16) & 0xff;
+  res[2] = (data.length >> 8) & 0xff;
+  res[3] = data.length & 0xff;
```

Comment:
Is the prefix length likely to ever change? If so, it would be more future-proof to make a const to hold "4" and then do the bit-twiddling in a for loop. If not, disregard.